src/tribler/gui/network/request.py

```diff
@@ -100,57 +100,64 @@ def _set_url(self, base_url: str):
                           for key, value in self.url_params.items()}
             self.url += '?' + urlencode(url_params, doseq=True)
 
-    def update_status(self, status_code: int):
-        self.logger.debug(f'Update {self}: {status_code}')
-        self.status_code = status_code
-        if status_code > 0:  # positive codes are HTTP response codes
-            self.status_text = str(status_code)
-        else:  # negative codes represent QNetworkReply error codes
-            self.status_text = f'{status_code}: {reply_errors.get(-status_code, "<unknown error code>")}'
-
     def on_finished(self):
         if not self.reply or not self.manager:
             return
 
         self.logger.info(f'Finished: {self}')
         try:
-            error_code = self.reply.error()
-            if error_code != QNetworkReply.NoError:
-                error_name = reply_errors.get(error_code, '<unknown error>')
-                self.logger.warning(f'Request {self} finished with error: {error_code} ({error_name})')
-                self.update_status(-error_code)
-                return
-
+            # If HTTP status code is available on the reply, we process that first.
+            # This is because self.reply.error() is not always QNetworkReply.NoError even if there is HTTP response.
+            # One example case is for HTTP Status Code 413 (HTTPRequestEntityTooLarge) for which QNetworkReply
+            # error code is QNetworkReply.UnknownContentError
```

Comment:
It doesn't look entirely correct to me to only analyze the HTTP response status code and no `QNetworkReply.NoError`. The HTTP status code is based on HTTP response headers. I believe it is possible that the client successfully received a `200 OK` header, but then the connection was closed prematurely while not all the response data was transferred. How does `QNetworkReply` handle this case? For normal responses that return JSON data, we can detect the incomplete response as `json.loads(data)` will fail. But for raw responses, we will probably get a truncated response from the server in that case.

But the case of a broken connection for raw response is probably seldom enough, so it may be OK not to handle it.

Reply:
If the HTTP response is 200, then the error code on QNetworkReply is 0, else it returns some other code. Here are some examples that I have tested myself.

HTTP--> QNetworkReply
200 --> 0
401 --> 204
400 --> 302
402 --> 299
403 --> 201
404 --> 203
413 --> 299
500 --> 401
502 --> 499

Since the error is non-zero for common HTTP status codes, I think the previous logic (where we checked the non-zero error code, logged the status, and returned) was incorrect and was masking the HTTP response body processing for common 4XX REST responses.

Regarding the case where the connection was closed prematurely while not all the response was transferred, I'm not sure how QNetworkReply handles it. However, I don't think we'd need to worry about such a case as the chances of that happening is negligible in my opinion.

Reply:
I agree that handling a broken connection, in this case, is not too important, as it should happen mostly during shutdown when GUI is already aware that Tribler is in shutdown state, and for JSON responses, the current try/except logic should handle incomplete responses without analyzing the QNetworkReply error codes. So I'm ok with this